### PR TITLE
a few changes related to LLVM support

### PR DESCRIPTION
**Pull request by smunix**
- refactor CMakeLists.txt
- add support for llvm@8
- segragate LLVMCompat libs for different versions of LLVM (WIP)
- initial addition of a basic ORC JIT compiler (WIP, llvm@6.0.1)

Request for comments...

### CODE REVIEW

File: bin/hi/main.C [hunk 1]
```diff
@@ -664,6 +665,8 @@ void initHI(evaluator* eval, bool useDefColors) {
 
 int main(int argc, char** argv) {
   try {
+    hobbes::orc::Jit::init();
```
Comment (kthielen) [inline bin/hi/main.C]: Why wouldn't `hobbes::cc` take care of this?

**Comment by kthielen** [comment]
This looks like an interesting direction!  The LLVM code has become pretty awkward/convoluted over the years as new versions have come and gone.  This looks like a path to factor that out and keep the version-specific logic in version-specific modules.

Could you comment on what the ORC part is about?  Is that for the new JIT lib that they've put in LLVM?

Also are you looking to factor this code to allow us to try other JIT options besides LLVM (like e.g. the basic/minimal header-only method pending review)?

File: bin/hi/main.C (hunk 1 above)
Comment (smunix) [reply bin/hi/main.C]: I'll definitely change this; this PR is giving a sense of where we're going with LLVM support in Hobbes, so as to gather early feedbacks.

**Comment by smunix** [comment]
> This looks like an interesting direction! The LLVM code has become pretty awkward/convoluted over the years as new versions have come and gone. This looks like a path to factor that out and keep the version-specific logic in version-specific modules.
> 

Yes, you got it right. 

> Could you comment on what the ORC part is about? Is that for the new JIT lib that they've put in LLVM?
> 

My hope is, I'll be able to factor out API that'd be generic enough to allow swapping in/out different implementations, other than LLVM. This part is still very much a WIP

> Also are you looking to factor this code to allow us to try other JIT options besides LLVM (like e.g. the basic/minimal header-only method pending review)?

Yep !

**Comment by kthielen** [comment]
Awesome!  This part of the code has been a little neglected over the past couple of years, good to see that it's getting some TLC.  :)

**Comment by kthielen** [comment]
Are we going to keep this?  Maybe consider another approach?

**Comment by smunix** [comment]
I'm no longer working on this approach, closing.

**Pull request closed without merging**